fpdf/drawing.py

```diff
@@ -0,0 +1,4207 @@
+from collections import OrderedDict
+from contextlib import contextmanager
+import copy
+import enum
+import decimal
+import math
+import re
+from typing import Optional, NamedTuple, Union
+
+__pdoc__ = {"force_nodocument": False}
+
+
+def force_nodocument(item):
+    """A decorator that forces pdoc not to document the decorated item (class or method)"""
+    __pdoc__[item.__qualname__] = False
+    return item
+
+
+@force_nodocument
+def force_document(item):
+    """A decorator that forces pdoc to document the decorated item (class or method)"""
+    __pdoc__[item.__qualname__] = True
+    return item
+
+
+# type alias:
+Number = Union[int, float, decimal.Decimal]
+NumberClass = (int, float, decimal.Decimal)
+
+
+# this maybe should live in fpdf.syntax
+class Raw(str):
+    """str subclass signifying raw data to be directly emitted to PDF without transformation."""
+
+
+class Name(str):
+    """str subclass signifying a PDF name, which are emitted differently than normal strings."""
+
+
+WHITESPACE = frozenset("\0\t\n\f\r ")
+"""Characters PDF considers to be whitespace."""
+EOL_CHARS = frozenset("\n\r")
+"""Characters PDF considers to mark the end of a line."""
+DELIMITERS = frozenset("()<>[]{}/%")
+"""Special delimiter characters"""
+
+
+NAME_ESC = re.compile(
+    b"[^" + bytes(v for v in range(33, 127) if v not in b"()<>[]{}/%#\\") + b"]"
+)
+STR_ESC = re.compile(r"[\n\r\t\b\f()\\]")
+STR_ESC_MAP = {
```

Comment:
Yes, it seems they accomplish much of the same purpose, though this one escapes more characters (following table 3.2 in the PDF 1.7 spec) but doesn't support bytes objects (mainly because my general PDF serialization function emits those as hexadecimal PDF strings). I was perhaps overzealous with encapsulating functionality in the drawing module, as I was trying to avoid breaking any existing functionality.

Arguably, this isn't necessarily more correct than the existing function. The section of the PDF spec that I've referenced doesn't specify how unescaped tabs/form feeds/backspaces are handled when encountered in PDF string literals. And it's not terribly clear to me, but from the wording, unescaped `\r\n` may be handled differently by the viewer than its escaped counterpart.

I'm not sure what the right call here is. I could easily scrap this and use `fpdf.utils.escape_parens` or I could update that function to do the additional escaping that this does.

Reply:
To me, your code is clearer so I'd go for solution 2 (update `pdf.utils.escape_parens` based on your escaping logic) and see if the non-regression tests pass. Id they don't, maybe opt for solution 1 then.
Also, it'd be nice to a comment mentioning table 3.2

Reply:
I made this change locally and it caused a number of tests to start failing on my computer. It seems that the failing tests are all ones that do something with a font, which is interesting. The diffs are a bit hard to interpret, but when I removed `\n` from the escape list, all the failing tests pass, so I guess it was due to escaping `\n`. I can commit this with `\n` removed from the escape list if you'd like.